src/coreclr/vm/tieredcompilation.cpp

```diff
@@ -355,88 +392,190 @@ void WINAPI TieredCompilationManager::TieringDelayTimerCallback(PVOID parameter,
 
-                if (!m_isBackgroundWorkScheduled && (m_isPendingCallCountingCompletion || m_countOfMethodsToOptimize != 0))
-                {
-                    m_isBackgroundWorkScheduled = true;
-                    doBackgroundWork = true;
-                }
+        if (IsTieringDelayActive())
+        {
+            do
+            {
+                ClrSleepEx(delayMs, false);
+            } while (!TryDeactivateTieringDelay());
+        }
 
-                break;
-            }
+        // Don't want to perform background work as soon as it is scheduled if there is possibly more important work that could
+        // be done. Some operating systems may also give a thread woken by a signal higher priority temporarily, which on a
+        // CPU-limited environment may lead to rejitting a method as soon as it's promoted, effectively in the foreground.
+        ClrSleepEx(0, false);
+
+        if (IsTieringDelayActive())
+        {
+            continue;
+        }
+
+        if ((m_isPendingCallCountingCompletion || m_countOfMethodsToOptimize != 0) &&
+            !DoBackgroundWork(&workDurationTicks, minWorkDurationTicks, maxWorkDurationTicks))
+        {
+            // Background work was interrupted due to the tiering delay being activated
+            _ASSERTE(IsTieringDelayActive());
+            continue;
         }
 
-        // Reschedule the timer if there has been recent tier 0 activity (when a new eligible method is called the first
-        // time) to further delay call counting
-        bool success = false;
-        EX_TRY
         {
-            if (ThreadpoolMgr::ChangeTimerQueueTimer(
-                    tieringDelayTimerHandle,
-                    g_pConfig->TieredCompilation_CallCountingDelayMs(),
-                    (DWORD)-1 /* Period, non-repeating */))
+            LockHolder tieredCompilationLockHolder;
+
+            if (IsTieringDelayActive() || m_isPendingCallCountingCompletion || m_countOfMethodsToOptimize != 0)
             {
-                success = true;
+                continue;
             }
+
+            s_isBackgroundWorkerProcessingWork = false;
         }
-        EX_CATCH
+
+        // Wait for the worker to be scheduled again
+        DWORD waitResult = s_backgroundWorkAvailableEvent.Wait(timeoutMs, false);
+        if (waitResult == WAIT_OBJECT_0)
         {
+            continue;
         }
-        EX_END_CATCH(RethrowTerminalExceptions);
-        if (success)
+        _ASSERTE(waitResult == WAIT_TIMEOUT);
+
+        // The wait timed out, see if the worker can exit
+
+        LockHolder tieredCompilationLockHolder;
+
+        if (s_isBackgroundWorkerProcessingWork)
         {
-            return;
+            // The background worker got scheduled again just as the wait timed out. The event would have been signaled just
+            // after the wait had timed out, so reset it and continue processing work.
+            s_backgroundWorkAvailableEvent.Reset();
```

Comment:
If we are going to manually reset the event in some cases is there a reason not to use a ManualResetEvent and Reset() it inside this lock always?

Reply:
Resetting manually would be more expensive than having it auto-reset. This reset is optional, it's a rare case where we know that the event is signaled unnecessarily and not doing the reset would cause the thread to loop an extra time, which would be more expensive unnecessarily. Auto-reset also feels more natural to me here, as in most cases, as it seems rare to me though valid, that one would want to unblock multiple waits on an event all at once.

Reply:
It is surprising to me to see AutoReset used as a perf-optimization to save the cost of the Reset() call in the common case. Given the asserts you needed to change in the event code I assume its a pattern not used anywhere else in the coreclr codebase. If the perf really matters then I don't mind going along but I would add more explanation in the comments about why it is written this way. I was quite puzzled by it : )

Reply:
> It is surprising to me to see AutoReset used as a perf-optimization to save the cost of the Reset() call in the common case.

I didn't use it as an optimization, I used it naturally because to me AutoReset is the natural thing to use here by default and there would need to be a good reason to use ManualReset instead (not only because it is less optimal).

Reply:
Also the assertion that AutoReset events shouldn't be manually Reset is completely arbitrary and unnecessary restriction, as shown here there are sometimes good reasons to manually reset an AutoReset event.

Reply:
I agree that it is unnecessary restriction, but the combination of my confusion, someone deliberately writing asserts to ensure it doesn't happen, and nobody else diverging from that pattern in the entire coreclr codebase is decent evidence that it is an established convention in place. Fwiw I would have found ManualResetEvent rather than AutoResetEvent to be the more natural primitive to use here but I only mention it to say I think its a matter of opinion. I view you as the owner of this code so I am fine to give your opinion the greater weight. My only request is to comment it because there is a decent chance others could find this confusing too.

Reply:
Sure I'll add a comment, but I don't think it should be necessary to add a comment every time one type of event is used over the other. If the type of event that is used works reasonably in the case where it is used and there is no particular logical reason to use the other type of event, then any established convention is maybe not very relevant for this type of thing. Conventions may be important for other things, I'm not aware of any strong convention regarding the type of event that is used, typically as far as I'm aware the type of event that makes sense to the author would be used, and unless there is a good reason to use one over the other that's totally fine.

Reply:
> I don't think it should be necessary to add a comment every time one type of event is used over the other

That isn't what I am proposing, perhaps I explained badly. I am suggesting:
- If the person doing the code review found something confusing, there is an increased chance that other developers would react similarly. I'm saying I found this specific instance of using an AutoResetEvent confusing.
- Asserts in the codebase are a good indication of what some other developer thought was a useful invariant. Code changes that delete those asserts so they can change the invariant have increased risk of creating some confusion. The assert in this case was that AutoResetEvent will never be used in a scenario where it needs to be manually reset.

Hopefully that sounds more sensible now : )

Reply:
The existence of the assert and because of it, the call to Reset, may have caused extra confusion. For the comment maybe for now just a minimal comment indicating that any type of event could be logically used here without going into too much detail, hopefully that would suffice to alleviate some confusion. Hopefully it shouldn't be necessary to indicate that in the future where either type of event could be used.